packages/freezed/lib/src/templates/concrete_template.dart

```diff
@@ -154,12 +156,32 @@ ${copyWith.abstractCopyWithGetter}
 
   String get _redirectedFromJsonConstructor {
     if (!shouldGenerateJson) return '';
-    return 'factory ${constructor.redirectedName}.fromJson(Map<String, dynamic> json) = $concreteName$genericsParameter.fromJson;';
+
+    final genericArgs = hasGenericArgumentFactories
+        ? genericsParameter.typeParameters.map((type) {
+            return ', $type Function(Object? json) fromJson$type';
+          }).join()
+        : '';
+
+    return 'factory ${constructor.redirectedName}.fromJson(Map<String, dynamic> json$genericArgs) = $concreteName$genericsParameter.fromJson;';
   }
 
   String get _concreteFromJsonConstructor {
     if (!shouldGenerateJson) return '';
-    return 'factory $concreteName.fromJson(Map<String, dynamic> json) => _\$${concreteName}FromJson(json);';
+
+    final genericArgs = hasGenericArgumentFactories
+        ? genericsParameter.typeParameters.map((type) {
+            return ', $type Function(Object? json) fromJson$type';
+          }).join()
+        : '';
+
+    final genericArgsNames = hasGenericArgumentFactories
+        ? genericsParameter.typeParameters
+            .map((type) => ', fromJson$type')
+            .join()
+        : '';
```

Comment:
same thing, make a reusable function

Reply:
In fact we probably should extract all the toJson logic in a separate template file

Reply:
I don't see where this is repeated

Reply:
It's repeated accross files.

As I said before, we'd want to extract this logic in a separate `templates/serialisation_template.dart` with all the fromJson/toJson logic.

Reply:
Done. Now all fromJson and toJson related logic is inside of `templates/serialization_template.dart`